paddlenlp/trainer/trainer.py

```diff
@@ -480,34 +480,41 @@ def recover_params_from_master_weights(self, state_dict):
     def _all_gather_state_dict(self, state_dict, filter_func, group=None):
         if group is None:
             group = self.hcg.get_sharding_parallel_group()
-        state_list = list(state_dict.items())
+        res = OrderedDict()
 
         def map_func(weight):
             if isinstance(weight, paddle.Tensor):
                 weight = weight.numpy()
             return weight
 
-        state_list = [(e[0], map_func(e[1])) for e in state_list]
-        len_res = []
-        obj_len = paddle.to_tensor(len(state_list))
-        gather_res = OrderedDict()
-        paddle.distributed.all_gather(len_res, obj_len, group=group)
-        max_len = int(max(len_res).item())
-        for i in range(max_len):
-            tmp = []
-            if i >= obj_len:
-                send_item = ("pack", np.array([]))
+        state_dict = {k: map_func(v) for (k, v) in state_dict.items()}
+
+        meta_dict = {}
+        for (k, v) in state_dict.items():
+            # src rank
+            meta_dict[k] = (v.dtype, v.shape, group.rank)
+
+        meta_dict_list = self._all_gather_simple_object(meta_dict, group)
+        meta_list = [(k, v) for meta in meta_dict_list for (k, v) in meta.items()]
+        meta_list = sorted(meta_list, key=lambda x: x[1])
+        for (k, meta) in meta_list:
+            dtype, shape, rank = meta
+            if rank == group.rank:
+                assert k in state_dict
+                tensor = paddle.to_tensor(state_dict[k])
             else:
-                send_item = state_list[i]
-            paddle.distributed.all_gather_object(tmp, send_item, group=group)
-            for recv_item in tmp:
-                if recv_item[0] == "pack" and recv_item[1].size == 0:
-                    continue
-                if not filter_func(recv_item[0]):
-                    continue
-                gather_res[recv_item[0]] = paddle.to_tensor(recv_item[1], place=paddle.CPUPlace())
-                print(f"{i} gather {recv_item[0]} ")
-        return gather_res
+                tensor = paddle.to_tensor(np.zeros(shape, dtype))
```

Comment:
Can use `tensor = paddle.empty(shape, dtype=dtype)` to avoid init CPU zero numpy array.